grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -1083,12 +1099,117 @@ class StockMovementService {
         List<PackPageItem> packPageItems = new ArrayList<PackPageItem>(items)
 
         if (max != null && offset != null) {
-            return packPageItems.subList(offset.toInteger(), offset.toInteger() + max.toInteger() > packPageItems.size() ? packPageItems.size() : offset.toInteger() + max.toInteger())
+            return packPageItems.subList(offset, offset + max > packPageItems.size() ? packPageItems.size() : offset + max)
         }
 
         return packPageItems
     }
 
+    List<ImportPickCommand> parsePickCsvTemplateImport(String text) {
+        try {
+            def csvMapReader = new CSVMapReader(new StringReader(text), [skipLines: 1])
+            csvMapReader.fieldKeys = [
+                    'id',
+                    'code',
+                    'name',
+                    'lot',
+                    'expiration',
+                    'binLocation',
+                    'quantity',
+            ]
+            List<Map> dataList = csvMapReader.toList()
+            return dataList*.asType(ImportPickCommand)
+
+        } catch (Exception e) {
+            throw new RuntimeException("Error parsing order item CSV: " + e.message, e)
+        }
+    }
+
+    void validatePicklistListImport(List<PickPageItem> pickPageItems, Boolean supportsOverPick, List<ImportPickCommand> picklistItems) {
+        Map<String, List> groupedItems = picklistItems.groupBy { it.id }
+
+        picklistItems?.each { ImportPickCommand data ->
+            data.validate()
+            // skip validation if id is empty since mos tof the validation relies on an existing requisition id
+            if (!data.id) {
+                return
+            }
+            PickPageItem pickPageItem = pickPageItems.find { it.requisitionItem?.id == data.id }
```

Comment:
I know you are already overwhelmed with this PR, so treat it just as a comment, but imho those manual validations could also be handled in the command's validator.
After this find, you could just set the `pickPageItem` to the `ImportPickCommand` and then call `.validate()` on the command.